File: BridgewellDemoObjectiveC/InAppBackgroundView.m
//
//  InAppBackgroundView.m
//  BridgewellDemoObjectiveC
//
//

#import "InAppBackgroundView.h"
#import <UIKit/UIKit.h>

@interface UIView (Inspectable)

@property (nonatomic, assign) IBInspectable CGFloat cornerRadius;

@end

@implementation UIView (Inspectable)

- (void)setCornerRadius:(CGFloat)cornerRadius {
    self.layer.cornerRadius = cornerRadius;
    self.clipsToBounds = YES;
}

- (CGFloat)cornerRadius {
    return self.layer.cornerRadius;
}

@end

@implementation InAppBackgroundView

@synthesize contentView = _contentView;

- (instancetype)initWithFrame:(CGRect)frame xibName:(NSString *)xibName {
    self = [super initWithFrame:frame];
    if (self) {
        self.xibName = xibName;
        [self xibSetup];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self xibSetup];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self xibSetup];
    }
    return self;
}

- (void)xibSetup {
    self.backgroundColor = [UIColor clearColor];
    _contentView = [self loadViewFromNib];
    _contentView.frame = self.bounds;
    [self addSubview:_contentView];
    _contentView.translatesAutoresizingMaskIntoConstraints = NO;

    [NSLayoutConstraint activateConstraints:@[
        [_contentView.topAnchor constraintEqualToAnchor:self.topAnchor],
        [_contentView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
        [_contentView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
        [_contentView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor]
    ]];

    self.translatesAutoresizingMaskIntoConstraints = NO;
}

- (UIView *)loadViewFromNib {
    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
    NSString *name = self.xibName != nil ? self.xibName : NSStringFromClass([self class]);
    return [[bundle loadNibNamed:name owner:self options:nil] firstObject];
}

@end

File: BridgewellDemo/InAppBackgroundView.swift
//
//  InAppBackgroundView.swift
//  ExampleApp
//
//

import Foundation
import UIKit

extension UIView {
    @IBInspectable var cornerRadius: CGFloat {
        set {
            layer.cornerRadius = newValue
            clipsToBounds = true
        }
        get {
            layer.cornerRadius
        }
    }
}

class InAppBackgroundView: UIView {
    
    private(set) lazy var contentView = loadViewFromNib()
    var xibName: String?
    
    init(frame: CGRect, xibName: String?) {
        super.init(frame: frame)
        self.xibName = xibName
        xibSetup()
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        xibSetup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        xibSetup()
    }
    
    func xibSetup() {
        backgroundColor = .clear
        contentView.frame = bounds
        addSubview(contentView)
        contentView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            contentView.topAnchor.constraint(equalTo: topAnchor),
            contentView.bottomAnchor.constraint(equalTo: bottomAnchor),
            contentView.leadingAnchor.constraint(equalTo: leadingAnchor),
            contentView.trailingAnchor.constraint(equalTo: trailingAnchor)
        ])
        translatesAutoresizingMaskIntoConstraints = false
    }
    
    private func loadViewFromNib() -> UIView {
        let bundle = Bundle(for: type(of: self))
        let name = xibName != nil ? xibName! : String(describing: type(of: self))
        return UINib(nibName: name, bundle: bundle).instantiate(withOwner: self, options: nil).first as! UIView
    }
    
}
